### PR TITLE
add support for typescript/babel transpilation for server process

### PR DESCRIPTION
**Pull request by ranyitz**
partially `fixes` #1210

### 🔦 Summary
This PR adds support for TypeScript/babel transpilation for server process file (e.g. `index-dev.ts`/`dev/server.ts`). It's now possible to write those files with the same syntax as the rest of the app.

* [x] TypeScript runs without failing on TypeChecking (`fast`) because it will happen in the IDE and that's not critical to fail the development build on type errors during `npm start` (which doesn't run in the CI)

* [x] Modify templates and remove the deprecated `--server` so they'll take `dev/server.ts` instead.

* [x] Add link to fixtures in contributing

### 🗺️ Test Plan
I've noticed we don't have a specific test for `app-server` but instead, all other integration tests are testing that it works. The easiest thing was to change `index-dev.js` to `index-dev.ts` in templates tests to verify that it works.

Added a test for the TypeScript case, where we test that the server doesn't fail on type errors.

(I've also tested it manually, and in client projects in templates)

### CODE REVIEW

File: packages/yoshi-common/src/server-process-worker.js [hunk 1]
```diff
@@ -0,0 +1,16 @@
+// We use this file in order to have a `fork` with `ipc`
+// and also have `ts-node` register for TypeScript support
+// for the user's server file
+
+const path = require('path');
+
+const serverFilePath = process.env.__SERVER_FILE__;
+
+if (path.extname(serverFilePath) === '.ts') {
```
Comment (ronami) [inline packages/yoshi-common/src/server-process-worker.js]: I think we should check for the existence of a `tsconfig.json` (`isTypeScriptProject()`, can be passed as an env var from outside), since some projects may use `allowJs` and have only some of their setup code in TypeScript.

File: packages/yoshi-common/src/server-process-worker.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+  require('ts-node').register({
+    fast: true,
+  });
+}
+
+// eslint-disable-next-line
```
Comment (ronami) [inline packages/yoshi-common/src/server-process-worker.js]: Why is that needed? Can we disable only a specific rule instead of all of them?

File: packages/yoshi-common/src/server-process-worker.js (hunk 1 above)
Comment (ranyitz) [reply packages/yoshi-common/src/server-process-worker.js]: Cool, so we have `ts-node` configured even if the initial file is a `js` file.

Good idea! I'll fix that

File: packages/yoshi-common/src/server-process-worker.js (hunk 2 above)
Comment (ranyitz) [reply packages/yoshi-common/src/server-process-worker.js]: It's because of the `require` call, i'll add the specific rule

Comment (ranyitz) [reply packages/yoshi-common/src/server-process-worker.js]: Actually it's not needed, my bad, removing

File: packages/yoshi-common/src/server-process-worker.js (hunk 1 above)
Comment (ronami) [reply packages/yoshi-common/src/server-process-worker.js]: Thinking about it again, maybe run `@babel/register`/`ts-node`, depending on the project type? We still support Babel and only having this feature for ts project will be annoying for existing Babel projects.

File: packages/yoshi-common/src/server-process-worker.js (hunk 2 above)
Comment (ronami) [reply packages/yoshi-common/src/server-process-worker.js]: I don't think a dynamic `require` call should show any linting errors since it cannot be done with an import. It's interesting to know what rule is failing.

File: packages/yoshi-common/src/server-process-worker.js (hunk 1 above)
Comment (ranyitz) [reply packages/yoshi-common/src/server-process-worker.js]: I've modified the code to handle both cases, I can easily add a test as well

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -39,7 +39,7 @@ Test each feature on a dedicated installed project.
 
 How it works:
 
-1. We copy base fixture template to `.tmp` folder (basic files like `package.json` etc.)
+1. We copy [base fixture template](./test/fixtures/typescript/base-template/package.json) to `.tmp` folder (file which are neede in most tests, like `package.json`)
```
Comment (ranyitz) [inline CONTRIBUTING.md]:
```suggestion
1. We copy [base fixture template](./test/fixtures/typescript/base-template/package.json) to `.tmp` folder (file which are needed in most tests, like `package.json`)
```

File: packages/yoshi-common/src/server-process-worker-with-transpilation.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+// We use this file in order to have a `fork` with `ipc`
+// We setup either `babel/register` or `ts-node`
+// for the user's server file to have transpilation enabled
+const userServerFilePath = process.env.__SERVER_FILE_PATH__;
+
+require('../build/require-hooks').setupRequireHooks();
```
Comment (ronami) [inline packages/yoshi-common/src/server-process-worker-with-transpilation.js]: I'm not sure if we need this since this file (`server-process-worker-with-transpilation.js`) will be in `/build` anyway. Can it be a TypeScript file with an import? That way, if we use it wrong or import something unknown we'll get a compilation error.

Comment (ranyitz) [reply packages/yoshi-common/src/server-process-worker-with-transpilation.js]: modified

**Comment by ranyitz** [comment]
@ronami Thanks for the great review! All code review comments have been fixed. Let me know if anything is missed.

File: packages/yoshi-common/src/server-process.ts [hunk 5]
```diff
@@ -57,7 +57,13 @@ export class ServerProcess {
       cwd: this.cwd,
     });
 
-    this.child = child_process.fork(this.serverFilePath, [], {
+    const userServerFilePath = path.join(process.cwd(), this.serverFilePath);
+    const serverProcessWorker = path.join(
+      __dirname,
+      'server-process-worker-with-transpilation.js',
+    );
```
Comment (ronami) [inline packages/yoshi-common/src/server-process.ts]: Super tiny (let's fix in a future commit) but can just be `require.resolve()`

File: test/typescript/features/app-server/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+    "lib": ["dom", "esnext", "webworker"],
+    "allowSyntheticDefaultImports": true,
+    "esModuleInterop": true,
+    "strict": true,
+    "resolveJsonModule": true,
+    "skipLibCheck": true
+  },
+  "files": [
+    "/Users/rany/Projects/yoshi/packages/yoshi/types.d.ts",
+    "/Users/rany/Projects/yoshi/node_modules/jest-yoshi-preset/types.d.ts"
```
Comment (ronami) [inline test/typescript/features/app-server/tsconfig.json]: Oops `/Users/rany/Projects/yoshi/packages/yoshi/types.d.ts`

Comment (ronami) [reply test/typescript/features/app-server/tsconfig.json]: Use `{%rootPath%}`

Comment (ranyitz) [reply test/typescript/features/app-server/tsconfig.json]: Opps 😨

**Comment by ranyitz** [comment]
@ronami Good points, new review comments fixed as well! Again, thanks for the review ❤️
